[Docs] Updating contributing guide and codespaces configuration (#741)
mckinsey/vizro: vizro-core/docs/pages/explanation/contributing.md

!!! note

    For either method, Hatch is the _only development dependency_. You do not need to manually install Python or create any virtual environments to develop Vizro; all this will be handled for you behind the scenes by Hatch. We have also configured our codespace to pre-install Hatch. If you develop on GitHub Codespaces you don't need to install anything at all!

## Develop on GitHub Codespaces

!!! note

    There is no need to manually create a fork of the Vizro code if you use GitHub Codespaces. A fork is [automatically created for you](https://docs.github.com/en/codespaces/developing-in-a-codespace/using-source-control-in-your-codespace#about-automatic-forking).

To develop on [GitHub Codespaces](https://docs.github.com/en/codespaces), follow the below steps:

1. [Create a codespace for our repository](https://codespaces.new/mckinsey/vizro). Leave the settings on their defaults and click "Create codespace" to start your codespace. It should take 1-2 minutes to fully launch and automatically start an example dashboard on port 8050.

> 2. Make changes to Vizro code in your codespace. See the [GitHub Codespaces documentation on developing in a codespace](https://docs.github.com/en/codespaces/developing-in-a-codespace/developing-in-a-codespace) for more information.

Reviewer: Do we want to suggest contributors create a feature branch in codespace, and changes go to the feature branch of their fork?

Reviewer 2: Good point, I also explain that in GHC as well. But on second thought this is probably mostly good practice for code hygiene - the PR will anyway come from another branch, so it is not necessary.

I am inclined to not mention it here because it keeps things simpler? Not entirely sure though

Author: I think it's fine for a PR from a fork to come from whatever branch they like. I'd only do a branch on a fork if I was working on multiple things in parallel. So I think fine to not mention here and just keep things simple.

3. Add your name to the [list of contributors](authors.md) (source file `vizro-core/docs/pages/explanation/authors.md`).

> 4. [Create a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork).

Reviewer: Tested with another account, the process is straightforward 👍 

When raising the PR, one check is failing: `Labeler / label (pull_request) `

Author: Good spot! I've just pushed something that will hopefully fix this but would be great to check again.

There are two ways to develop on Vizro:

* [GitHub Codespaces](https://docs.github.com/en/codespaces). This is the recommended method if you are a new contributor and is the quickest and easiest way to get started. All development can be done in your browser in a temporary environment; you do not need to set up anything on your computer. Read [Develop on GitHub Codespaces](#develop-on-github-codespaces) for full instructions on how to do this.
* Local machine. If you are more experienced then you might prefer to develop on your own computer. Read [Develop locally](#develop-locally) for full instructions on how to do this.

!!! note

    For either method, Hatch is the _only development dependency_. You do not need to manually install Python or create any virtual environments to develop Vizro; all this will be handled for you behind the scenes by Hatch. We have also configured our codespace to pre-install Hatch. If you develop on GitHub Codespaces you don't need to install anything at all!

## Develop on GitHub Codespaces

!!! note

>     There is no need to manually create a fork of the Vizro code if you use GitHub Codespaces, because fork is [automatically created for you](https://docs.github.com/en/codespaces/developing-in-a-codespace/using-source-control-in-your-codespace#about-automatic-forking).

Reviewer 2: ```suggestion
    There is no need to manually create a fork of the Vizro code if you use GitHub Codespaces, because a fork is [automatically created for you](https://docs.github.com/en/codespaces/developing-in-a-codespace/using-source-control-in-your-codespace#about-automatic-forking).
```

As I said in catchup - this actually was not 100% without friction. It did work, but there was an error message.

Author: Indeed... I would be tempted to leave these instructions as they are here (assuming the automatic fork works) and then you show the other flow where you do a manual fork in your GHC. wdyt?

Reviewer 2: Hm ok - let's do that. Do you then think it is an issue that I even show something different at GHC - I just don't want people to even encounter any error messages

Reviewer 2: As mentioned above, I am now not creating a fork first for GHC, but let's be aware that there sometimes can be issues..

    
To develop on [GitHub Codespaces](https://docs.github.com/en/codespaces), follow the below steps:

> 1. Use [Vizro's GitHub issues](https://github.com/mckinsey/vizro/issues) to discuss your contribution.

Reviewer 3: Wondering if this should be pulled out to the top to explain clearly how to find a ticket to work on and what to do before you get started if the work is more than a small tweak?

Reviewer 3: Maybe it's as simple as putting this big back that you removed (with some edit to add a bit about the requirements of a PR, which I've commented just below here)?

> Contributions of all experience levels are welcome! There are many ways to contribute, and we appreciate any help. Use our [issues page](https://github.com/mckinsey/vizro/issues) to discuss any contributions. Before opening a pull request, ensure you've first opened an issue to discuss the contribution.

Reviewer 3: Also, I think it would help to edit slightly, here's my suggestion for the first sentence:

```suggestion
Contributions of all experience levels are welcome! There are many ways to contribute, and we appreciate any help: it doesn't have to be a pull request on our code. You can also raise an issue to suggest an improvement to Vizro or our documentation, request a feature, or ask and answer community questions. See the FAQ for guidance on how to do these things.

<Then continue about the issues page blah blah>
```

Reviewer 2: @stichbury Do we use punctuation at the end of itemized lists?

Reviewer 3: It depends. Typically we do if the bullets comprise one or more full sentence, but not if they're not because they're just statements. So it's like this pair of examples...

Some of the features include:
* feature 1
* feature 2
* feature 3

You need to consider the following:
* Do you really want to hurt me?
* Do you really want to make me cry?
* Do you know any other song lyrics that are questions?

(but it doesn't need to be a question, could just be a period as long as the text is a full and complete sentence).

HOWEVER, I don't always adhere to this and it can be difficult because sometimes bullets are not always consistent in the list itself (which could make it look weird) and sometimes there's another bulleted list nearby on the page that should be punctuated so it looks odd if there's one that is and one that isn't. 

TL;DR: I'd typically say, go with punctuation more times than not, and if in doubt, add it.

Author: This is the sort of guide I was trying to follow so good to know, thanks @stichbury!

And good suggestions with these edits, I will definitely change as you suggest.

2. [Create a codespace for our repository](https://codespaces.new/mckinsey/vizro). Leave the settings on their defaults and click "Create codespace" to start your codespace. It should take 1-2 minutes to fully launch and automatically start an example dashboard on port 8050.
3. Make changes to Vizro code in your codespace. See the [GitHub Codespaces documentation on developing in a codespace](https://docs.github.com/en/codespaces/developing-in-a-codespace/developing-in-a-codespace) for more information.
4. Add your name to the [list of contributors](authors.md) (source file `vizro-core/docs/pages/explanation/authors.md`).
5. [Create a pull request (PR)](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork).
6. Before it can be merged, your PR must meet all the following requirements:
    * Two approving reviews (including a code owner).
    * Continuous Integration (CI) checks pass.

>     * Up-to-date with `main`.

Reviewer 3: I'm tempted to take this out and put it up top so it applies to both methods of contributing, to avoid repetition.

Also, should you add something about the changelog requirement here too, and link through to the hatch command to add it?

Reviewer 2: Should we also mention that some CI needs to be activated?

Author: @stichbury yeah, I wasn't sure about this. Given that a contributor would be using either the Codespace route or the local development route I thought it would be easier for someone to just follow all the steps in one of those sections without having to jump between a separate section that contains the parts that are common to both those processes and then back again. But it definitely makes sense to pull out point 1 on the list of both to the top of the page, so let me try and move this last point there as well and see if it reads ok.

I decided to not add anything about the changelog requirement here because:
* it's included in the "Continuous Integration (CI) checks pass" point
* when that check fails, the error message is very clear and easy to follow since it instructs the user to run `hatch run changelog:add`
* in theory changelogs aren't needed for every change anyway (purely docs changes shouldn't require it - if that's not the case then it's not working right)

@maxschulz-COL what do you mean by activated? Do the required jobs for external contributors not get triggered without some intervention?

Reviewer 2: I was under the impression that yes. When I did my testing I wasn't focussing on that, but it looked to me like it. @l0uden do you know if there is any activation needed when raising a PR from a fork?

I can test later again though just to be sure

Author: 1. Indeed seems like the automatic forking is not 100% reliable... but if it's working for you now then agreed with your proposal - let's leave the contribution guide as is then 👍 
2. Ah yes, the reason for this is this setting (which we currently have to be as permissive as possible). I will add a mention to this to the guide, good spot!
![image](https://github.com/user-attachments/assets/d1b3b18f-c350-44a7-b975-b5a14d72592e)

Author: @maxschulz-COL this is well worth knowing for GHC actually, given that many of them might have new github accounts so fit into this category.

3. [Fork the Vizro repository](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo) and clone it to your local machine.
4. Make changes to Vizro code in your fork.
5. Add your name to the [list of contributors](authors.md) (source file `vizro-core/docs/pages/explanation/authors.md`).
6. [Create a pull request (PR)](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork).
7. Before it can be merged, your PR must meet all the following requirements:
    * Two approving reviews (including a code owner).
    * Continuous Integration (CI) checks pass.
    * Your fork is up-to-date with `main`.

## How to use Hatch

Regardless of whether you are developing locally or in a codespace, everything you need to develop on Vizro is provided by Hatch through the [`hatch run`](https://hatch.pypa.io/latest/cli/reference/#hatch-run) command. The first time you use this command it will install all the required dependencies, including Python, which may take some time.

> The Hatch commands you need most commonly are as follows. These must be executed with `vizro-core` as your current working directory:

Reviewer 2: ```suggestion
The Hatch commands you need most commonly are as follows. These must be executed with `vizro-core` as your current working directory (or `vizro-ai` if your are working on our that package):
```

### `hatch run changelog:add`

`hatch run changelog:add` generates a new changelog fragment. This is included in CI checks and required for all changes outside documentation.

The format of our changelog is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). We use [scriv](https://pypi.org/project/scriv/) to build and maintain [our changelog](https://github.com/mckinsey/vizro/blob/main/vizro-core/CHANGELOG.md). When raising a PR, you must ensure that a changelog fragment has been created. This fragment is a small `.md` file describing your changes.

Run `hatch run changelog:add` to create a changelog fragment and then uncomment the relevant section(s). If you are uncertain about what to add or whether to add anything at all, refer to [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). The rule of thumb is that if Vizro users would be affected in any way then the changes should be described in the changelog.

!!! note

>     Changes that only affect documentation do not need a changelog fragment. This facilitates simple modifications to documentation [made directly on GitHub](https://docs.github.com/en/repositories/working-with-files/managing-files/editing-files) or with the [github.dev](https://docs.github.com/en/codespaces/the-githubdev-web-based-editor), where no terminal is available to run `hatch changelog:add`. Any changes outside documentation require a changelog fragment to be generated. If your changes do not require a changelog entry then you still need to generate the fragment but can leave it all commented out.

Reviewer 3: I have never found it to be the case that I can get away with no changelog, for any docs change. Is this a recent improvement?

Author: Not that recent: https://github.com/mckinsey/vizro/pull/480.

I did test it pretty thoroughly at the time and thought it seemed to work well, so if it's not working for you then please do point me towards a PR when you next find that it's not working as intended!

### `hatch run test-unit`

`hatch run test-unit` runs the test suite. This is included in CI checks.

Tests are handled using [pytest](https://docs.pytest.org/) and arguments are passed through to the underlying `pytest` command. For example, to rerun only failures from the last `pytest` invocation, you could run:

```console
hatch run test-unit --last-failed
```

In CI, we test across multiple Python versions and also [check for code coverage](https://coverage.readthedocs.io/). If required, you can also run this locally. For example, to run unit tests with Python 3.10 and check for code coverage, you would run:

```console
hatch run all.py3.10:test-unit-coverage
```

> In addition to unit tests with code coverage, CI also performs the following checks:

Reviewer 2: ```suggestion
In addition to run unit tests with code coverage, CI also performs the following checks:
```

> * `hatch run test-integration` runs integration tests that includes checking that the example apps in `vizro-core/examples` run.

Reviewer 2: ```suggestion
* `hatch run test-integration` runs integration tests that include checking that the example apps in `vizro-core/examples` run.
```

Vizro's dependencies are described by the `dependencies` section in `vizro-core/pyproject.toml`. There is no need to manually install or update the dependencies in your environment; they will be handled automatically for you when you do `hatch run`. This means that there is generally no need to `pip install` anything.

We have [configured Hatch to use uv](https://hatch.pypa.io/1.12/how-to/environment/select-installer/) for virtual environment creation, dependency resolution and installation. This is extremely fast. If you have installed unwanted dependencies in your Hatch environment then the simplest solution is to delete the environment (`hatch env remove` to remove one environment or `hatch env prune` to remove all environments). Your next `hatch run` command will quickly recreate the environment and install all the dependencies it needs.

If for some reason you do need to use `pip` then the correct way to do so is through `hatch run pip`. For example, you could run `hatch run pip show plotly`. This will use the version of uv that Hatch itself uses under the hood. If you already have uv installed globally then `uv pip show plotly` would also work.

> You _should not_ try to interact with Vizro dependencies using a global `pip`. For example, running `pip show plotly` without the `hatch run` prefix will not work correctly.

Reviewer 2: ```suggestion
!!! warning
    You _should not_ try to interact with Vizro dependencies using a global `pip`. For example, running `pip show 
    plotly` without the `hatch run` prefix will not work correctly.
```

1. Use [Vizro's GitHub issues](https://github.com/mckinsey/vizro/issues) to discuss your contribution.
2. [Create a codespace for our repository](https://codespaces.new/mckinsey/vizro). Leave the settings on their defaults and click "Create codespace" to start your codespace. It should take 1-2 minutes to fully launch and automatically start an example dashboard on port 8050.
3. Make changes to Vizro code in your codespace. See the [GitHub Codespaces documentation on developing in a codespace](https://docs.github.com/en/codespaces/developing-in-a-codespace/developing-in-a-codespace) for more information.
4. Add your name to the [list of contributors](authors.md) (source file `vizro-core/docs/pages/explanation/authors.md`).
5. [Create a pull request (PR)](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork).
6. Before it can be merged, your PR must meet all the following requirements:
    * Two approving reviews (including a code owner).
    * Continuous Integration (CI) checks pass.

>     * Up-to-date with `main`.

Author: ```suggestion
    * Code is up-to-date with `main`.
```

## Develop locally

1. Use our [GitHub issues](https://github.com/mckinsey/vizro/issues) to discuss your contribution.
2. Install Hatch. There are [several ways to do this](https://hatch.pypa.io/latest/install/).
3. [Fork the Vizro repository](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo) and clone it to your local machine.
4. Make changes to Vizro code in your fork.
5. Add your name to the [list of contributors](authors.md) (source file `vizro-core/docs/pages/explanation/authors.md`).
6. [Create a pull request (PR)](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork).
7. Before it can be merged, your PR must meet all the following requirements:
    * Two approving reviews (including a code owner).
    * Continuous Integration (CI) checks pass.

>     * Your fork is up-to-date with `main`.

Author: ```suggestion
    * Code is up-to-date with `main`.
```

## How to use Hatch

> Regardless of whether you are developing locally or in a codespace, everything you need to develop on Vizro is provided by Hatch through the [`hatch run`](https://hatch.pypa.io/latest/cli/reference/#hatch-run) command. The first time you use this command it will install all the required dependencies, including Python, which may take some time.

Author: ```suggestion
Regardless of whether you are developing locally or in a codespace, everything you need to develop on Vizro is provided by Hatch through the [`hatch run`](https://hatch.pypa.io/latest/cli/reference/#hatch-run) command. The first time you use this command it will install all the required dependencies, including Python.
```

Author: FYI @stichbury thanks to uv this is actually now ridiculously fast (like <30s) 😀 The only slow part of our local development process now is initially running linting.

### `hatch run docs:serve`

> `hatch run docs:serve` builds and displays documentation that is hot-reloading for convenience while you edit it. Documentation is also built automatically in your PR and can be previewed on Read The Docs. To do this, scroll to the bottom of your PR where all the checks are listed and click the "Details" link next to the Read the Docs build.

Author: ```suggestion
`hatch run docs:serve` builds and displays documentation that hot-reloads while you edit it. Documentation is also built automatically in your PR and can be previewed on Read The Docs. To do this, scroll to the bottom of your PR where all the checks are listed and click the "Details" link next to the Read the Docs build.
```

For more information on our documentation style, refer to our [style guide](documentation-style-guide.md).

### `hatch run pip`

> `hatch run pip` provides a [pip-compatible interface using uv](https://docs.astral.sh/uv/pip/). You should not need to use this.

Author: ```suggestion
`hatch run pip` provides a [pip-compatible interface using uv](https://docs.astral.sh/uv/pip/). You should not need to use this often.
```
